\section{Introduction}
\subsection{Motivation}
The polyadenylation signal(PAS) is part of the 3' untranslated region(UTR) of almost every messenger RNA. It is the binding site for a protein complex that causes the formation of the polyadenylation tail(poly(A)-tail). This tail of adenosines prevents the degradation of the mRNA in the cytoplasm effectively affecting the half-life of a mRNA and thereby raising the chance to be translated into a protein by the ribosome. As such, variants that have an impact on the efficiency of the PAS are potentially pathogenic due to their interference with the gene expression. Chen et al. wrote a review about diseases caused by PAS-affecting mutations. [REFERENCE] 

Prediction of polyadenylation sites has been attempted several times with moderate success in the last twenty years. [REFERENCES/EXAMPLES] Seeing that we cannot predict PASs with high accuracy prompts the question if we have enough knowledge about the structure of polyadenylation sites. Analyzing the secondary structure of the 3' UTR might give us more insight and allows the use of more characteristics for the prediction.

Accurately identifying functional PASs allows for better detection of disease causing mutations.  
%Kann zu Polyadenylation:
%[The PAS is a well-conserved six nucleotide long motif located usually ten to thirty bases before the cleavage site. [REFERENCE] Knowing this, it is possible to detect a PAS by using a naive search in a mRNA database. However, there are several problems with this approach. Assuming all four bases are evenly distributed throughout the DNA the same hexamer appears about every 4000 bases making the motif sequence a common occurrence.] 
\subsection{Background}
\subsubsection{History}
Polyadenylation (poly(A)) is the process of adding adenosines to the 3' end of precursor mRNA (pre-mRNA), i.e. messenger RNA that has been recently transcribed and has yet to undergo any post-transcription steps like splicing or polyadenylation.

The enzyme polyadenylate polymerase which catalyzes the reaction was first discovered in 1960. \cite{pmid13819354} But the protein's function was revealed about a decade later. \cite{pmid5288383} At that time it was still a mystery why most mRNAs had polyadenylated tails. 

In the early 1980s the role of the poly(A)-tail was largely elucidated: nuclear export, translation and stability of the mature mRNA. (reviewed in \cite{pmid6111419}). 

It took another decade to discover the two primary proteins that mediate the 3' end cleavage and polyadenylation, namely cleavage and polyadenylation specificity factor (CPSF) and cleavage stimulation factor (CStF). The number of participating proteins has since increased and recent publications mention several dozen cis-acting and auxiliary proteins or polypeptides working together in a large complex. \cite{pmid23774734} 

2008 it became clear that most genes have more than one polyadenylation site, resulting in transcripts with different lengths but identical coding sequence. \cite{pmid18411206} The new finding was termed alternative polyadenylation (APA). Since the 3' untranslated Region (3' UTR) has many binding sites for mRNA-repressive micro RNAs (miRNA), APA is taking part in the gene expression. \cite{pmid18566288}
\subsection{messenger ribonucleic acid (mRNA)}
RNA is a molecule present in all lifeforms. There are many types of RNA like transfer RNA, ribosomal RNA, microRNA or messenger RNA. Most RNA types play key roles in coding (mRNA), decoding (tRNA), regulation (microRNA) and expression of genes.

mRNA is particularly interesting because it transports genetic information stored by the DNA to the protein synthesis factories of a cell, the ribosome. There, the mRNA specifies the amino acids sequence of a protein, making it an important component of the gene expression.

Every mRNA is composed of three parts: the 5' untranslated region (5' UTR), the coding sequence (CDS) and the 3' untranslated region (3' UTR). [PICTURE] 

The 5' UTR is described as the region upstream of the start codon. It contains structures to regulate the translation of the transcript, like the 5' guanine cap. Although, it is called UTR, the 5' UTR may contain upstream open reading frames translating to peptides that inhibit the translation of the CDS of the transcript.

The CDS contains the sequence that is translated into an amino acid polymer by the ribosome.

The 3' UTR is defined as the region downstream of the stop codon. (see \ref{3_UTR})

mRNA is generated by the RNA polymerase II protein during the transcription process of a protein-coding gene. While prokaryotic and archaeal mRNA are rarely modified, eukaryotic mRNA is usually subject to many processing steps, including the addition of a 5' cap, splicing and polyadenylation. To distinguish between nascent and processed mRNA, the literature has termed them precursor mRNA and mature mRNA correspondingly.

The 5' cap is a special guanine that is added co-transcriptionally to the 5' end of a mRNA. It is needed for the transport of the mRNA from the nucleus to the cytoplasm and prevents the degradation by exonucleases. The cap is also important for the initiation of the translation in the ribosome. 

Splicing takes place during or after transcription and is responsible for the removal of the introns and the joining of the exons of most eukaryotic pre-mRNAs. Interestingly, splicing can produce different mRNAs. 

Polyadenylation occurrs right after the transcription terminates and describes the process of adding adenosines at the 3' end of the mRNA. (see \ref{polyadenylation})

\subsubsection{Polyadenylation}
\label{polyadenylation}
\subsubsection{3' UTR}
\label{3_UTR}
